ads/google/utils.js

```diff
@@ -14,16 +14,280 @@
  * limitations under the License.
  */
 
+import {documentStateFor} from '../../src/document-state';
+import {getService} from '../../src/service';
+import {timer} from '../../src/timer';
+import {
+  validateExperimentIds,
+  parseExperimentIds,
+} from './traffic-experiments';
+
+
 /**
  * @param {!Window} global
  * @return {number}
  */
 export function getCorrelator(global) {
-  const clientId = global.context.clientId;
-  const pageViewId = global.context.pageViewId;
-  if (global.context.clientId) {
+  return getAmpCorrelator(global.context.clientId, global.context.pageViewId);
+}
+
+/**
+ * @param {?string} clientId
+ * @param {string} pageViewId
+ * @return {number}
+ */
+export function getAmpCorrelator(clientId, pageViewId) {
+  if (clientId) {
     return pageViewId + (clientId.replace(/\D/g, '') % 1e6) * 1e6;
   } else {
     return pageViewId;
   }
 }
+
+/**
+ * @param {number} time
+ * @param {number} start
+ * @return {number|string}
+ */
+function elapsedTimeWithCeiling(time, start) {
+  const duration = time - start;
+  if (duration >= 1e6) {
+    return 'M';
+  } else if (duration >= 0) {
+    return duration;
+  }
+  return '-M';
+}
+
+class AdsenseInfo {
+  constructor() {
+    /** @private {number} */
+    this.nextSlotNumber_ = 0;
+  }
+  /**
+   * @return {number}
+   */
+  nextSlotNumber() {
+    return this.nextSlotNumber_++;
+  }
+}
+
+export function getAdsenseInfo(win) {
+  return getService(win, 'adsenseInfo', () => new AdsenseInfo());
+}
+
+const SIZE_EXCEPTION_ID = -12245933;
+
+/**
+ * @param {number} num Number, non-negative.
+ * @param {number} digits Number of digits, max 20.
+ * @return {string}
+ */
+function formatFixedWidthInteger(num, digits) {
+  const intPart = String(Math.max(Math.round(num), 0));
+  const len = intPart.length;
+  digits = Math.min(digits, 20);
+  if (len > digits) {
+    return '99999999999999999999'.substr(0, digits);
+  }
+  return '00000000000000000000'.substr(0, digits - len) + intPart;
+}
+
+function adKey(slotNumber, slot, viewport, screen) {
+  // Assume Nexus 6 screen, if we don't get either a viewport or a screen.
+  const viewWidth = viewport ? viewport.width : screen ? screen.width : 1080;
+  const viewHeight = viewport ? viewport.height : screen ? screen.height : 1920;
+  const slotTop = slot.y + (viewport ? viewport.y : 0);
+  const slotLeft = slot.x + (viewport ? viewport.x : 0);
+  return formatFixedWidthInteger(slotNumber, 2) +
+    // ad slot top, in 1/5 viewport height units
+    formatFixedWidthInteger(slotTop * 5 / viewHeight, 4) +
+    // ad slot left, in 1/5 viewport width units
+    formatFixedWidthInteger(slotLeft * 5 / viewWidth, 4);
+}
+
+/**
+ * Make the URL for an ad request, from an amp-ad element.
+ * @param {number} slotNumber
+ * @param {!Window} global
+ * @param {!Object} data
+ * @param {!IntersectionObserverEntry} intersectionRecord
+ * @return {string}
+ */
+export function adsenseRequestURLForAmpAd(slotNumber, global, data,
```

Comment:
super nit that I don't feel strongly about: In Google JS style guide it would be `Url`